XMonad/Hooks/BorderPerWindow.hs

```diff
@@ -0,0 +1,71 @@
+-----------------------------------------------------------------------------
+-- |
+-- Module      :  XMonad.Hooks.BorderPerWindow
+-- Description :  Set border width for each window in all layouts.
+-- Copyright   :  (c) 2021 Xiaokui Shu
+-- License     :  BSD-style (see LICENSE)
+--
+-- Maintainer  :  subbyte@gmail.com
+-- Stability   :  unstable
+-- Portability :  unportable
+--
+-- Want to customize border width, for each window on all layouts? Want
+-- specific window have no border on all layouts? Try this.
+-----------------------------------------------------------------------------
+
+module XMonad.Hooks.BorderPerWindow ( -- * Usage
+                                      -- $usage
+
+                                      -- * Design Considerations
+                                      -- $design
+                                      defineWindowBorder
+                                    , setCustomWindowBorder
+                                    ) where
+
+
+import XMonad
+import qualified XMonad.Util.ExtensibleState as XS
+import Graphics.X11.Xlib (Dimension)
+import XMonad.Util.ActionQueue (enqueue, exequeue)
+
+-- $usage
+--
+-- 1. Specify which window to customize border in 'manageHook' of the config:
+--
+-- > myManageHook :: ManageHook
+-- > myManageHook = composeAll
+-- >     [ className =? "firefox"  --> defineWindowBorder 0
+-- >     , className =? "Chromium" --> defineWindowBorder 0
+-- >     , isDialog                --> defineWindowBorder 8
+-- >     ]
+--
+-- 2. Add the two hooks in your config:
+--
+-- > myConfig = def
+-- >     { ...
+-- >     , logHook = setCustomWindowBorder
+-- >     , manageHook = myManageHook
+-- >     , ...
+-- >     }
+
+-- $design
+--
+-- 1. Keep it simple. Since the extension does not aim to change border setting
+--    when layout changes, only execute the border setting function once to
+--    avoid potential window flashing/jumping/scaling.
+--
+-- 2. The 'ManageHook' eDSL is a nice language for specifying windows. Let's
+--    build on top of it and use it to specify window to define border.
+
+defineBorderWidth :: Dimension -> ManageHook
+defineBorderWidth bw = do
+    w <- ask
+    liftX . XS.modify . enqueue $ updateBorderWidth w bw
+    idHook
+
+updateBorderWidth :: Window -> Dimension -> X ()
+updateBorderWidth w bw = do
+    withDisplay $ \d -> io $ setWindowBorderWidth d w bw
+    refresh
+
+setCustomWindowBorder = exequeue
```

Comment:
I think I would be against an alias like this.  If another module decides to use X.A.ActionQueue also does this you'd have the same logHook twice in there; that doesn't feel right

Reply:
Perhaps the ActionQueue module could export just an `XConfig a → XConfig a` combinator which configures the logHook exactly once (using extensibleConf)?

Reply:
(I'll be happy to refactor this later, which I was going to do anyway as I think the queue handling needs some improvements but I can't look into it today and don't want to block the rest of you.)

Reply:
> Perhaps the ActionQueue module could export just an `XConfig a → XConfig a` combinator which configures the logHook exactly once (using extensibleConf)?

Oh, right, we have tools for that now!  In that case this sounds like the right thing to do

> (I'll be happy to refactor this later, which I was going to do anyway as I think the queue handling needs some improvements but I can't look into it today and don't want to block the rest of you.)

I don't suppose there is much of a hurry (except for @subbyte not having to run their own fork anymore :)); waiting a few days longer seems very much acceptable